src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/CreateOrUpdateThirdPartyConfigAPI.java

```diff
@@ -0,0 +1,114 @@
+import java.util.List;
+
+public class CreateOrUpdateThirdPartyConfigAPI extends WebserverAPI {
+
+    private static final long serialVersionUID = -4641988458637882374L;
+
+    public CreateOrUpdateThirdPartyConfigAPI(Main main) {
+        super(main, RECIPE_ID.MULTITENANCY.toString());
+    }
+
+    @Override
+    public String getPath() {
+        return "/recipe/multitenancy/config/thirdparty";
+    }
+
+    @Override
+    protected void doPut(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        String thirdPartyId = InputParser.parseStringOrThrowError(input, "thirdPartyId", false);
+
+        try {
+            TenantIdentifierWithStorage tenantIdentifier = this.getTenantIdentifierWithStorageFromRequest(req);
+            TenantConfig config = Multitenancy.getTenantInfo(main, tenantIdentifier);
+            if (config == null) {
+                throw new TenantOrAppNotFoundException(tenantIdentifier);
+            }
+            List<ThirdPartyConfig.Provider> newProviders = new ArrayList<>();
+
+            boolean found = false;
+            for (ThirdPartyConfig.Provider provider: config.thirdPartyConfig.providers) {
+
+                if (!provider.thirdPartyId.equals(thirdPartyId)) {
+                    newProviders.add(provider);
+                } else {
+                    ThirdPartyConfig.Provider newProvider = new Gson().fromJson(input, ThirdPartyConfig.Provider.class);
+                    newProviders.add(newProvider);
+                    found = true;
+                }
+            }
+            if (!found) {
+                ThirdPartyConfig.Provider newProvider = new Gson().fromJson(input, ThirdPartyConfig.Provider.class);
+                newProviders.add(newProvider);
+            }
+            TenantConfig updatedConfig = new TenantConfig(
+                    config.tenantIdentifier,
+                    config.emailPasswordConfig,
+                    new ThirdPartyConfig(
+                            config.thirdPartyConfig.enabled, newProviders.toArray(new ThirdPartyConfig.Provider[0])),
+                    config.passwordlessConfig,
+                    config.coreConfig);
+
+            TenantIdentifier sourceTenant;
+            if (!tenantIdentifier.getTenantId().equals(TenantIdentifier.DEFAULT_TENANT_ID)) {
+                sourceTenant = new TenantIdentifier(
+                        tenantIdentifier.getConnectionUriDomain(), tenantIdentifier.getAppId(), null);
+            } else if (!tenantIdentifier.getAppId().equals(TenantIdentifier.DEFAULT_APP_ID)) {
+                sourceTenant = new TenantIdentifier(
+                        tenantIdentifier.getConnectionUriDomain(), null, null);
+            } else {
+                sourceTenant = new TenantIdentifier(null, null, null);
+            }
+            Multitenancy.addNewOrUpdateAppOrTenant(main, sourceTenant, updatedConfig);
```

Comment:
remove sourceTenant from addNewOrUpdateAppOrTenant and do the permission check elsewhere (in the API layer)